index.d.ts

```diff
@@ -666,51 +1237,133 @@ interface DurableObjectOperator {
   list<T = unknown>(options?: DurableObjectListOptions): Promise<Map<string, T>>;
 }
 
+// TODO(detail)
 interface DurableObjectTransaction extends DurableObjectOperator {
   rollback(): void;
 }
 
+// TODO(detail)
 interface DurableObjectStorage extends DurableObjectOperator {
   transaction(
     closure: (txn: DurableObjectTransaction) => Promise<void>
   ): Promise<void>;
 }
 
-interface DurableObjectState {
-  id: DurableObjectId;
-  storage: DurableObjectStorage;
+/**
+ * @deprecated Use `DurableObjectContext` instead.
+ */
+type DurableObjectState = DurableObjectContext;
+
+/**
+ * The Durable Object context.
+ */
+interface DurableObjectContext extends Context {
+  /**
+   * The object ID.
+   */
+  readonly id: DurableObjectId;
   /**
-   * Use this method to notify the runtime to wait for asynchronous tasks
-   * (e.g. logging, analytics to third-party services, streaming and caching).
+   * The persistent object storage.
    */
-  waitUntil(promise: Promise<any>): void;
+  readonly storage: DurableObjectStorage;
 }
 
 /**
- * DurableObject is a class that defines a template for creating Durable Objects
+ * A `DurableObject` is a class that defines the template for each
+ * Durable Object.
  */
-interface DurableObject {
-  fetch: (request: Request) => Promise<Response>;
+declare abstract class DurableObject<Environment = unknown> {
```

Comment:
This doesnt work. While it may be correct on its own, abstract classes serve no use to user-facing code because it would require they do this:

```ts
class Counter extends DurableObject<Bindings> {
  // Have to redeclare all type signatures again anyway
  constructor(ctx: DurableObjectContext, env: Bindings) {
    super(ctx, env); // have to super for TS, but is invalid runtime
  }
```

There's unfortunately no way around needing to redeclare all types for class properties and/or methods.
I say "this doesn't work" specifically because of the `abstract class` type which will force people to add that `super` code to satisfy TS, but will throw a runtime error.